services/repository/license.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2024 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"code.gitea.io/gitea/models/db"
+	repo_model "code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/modules/container"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/gitrepo"
+	"code.gitea.io/gitea/modules/graceful"
+	"code.gitea.io/gitea/modules/json"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/options"
+	"code.gitea.io/gitea/modules/queue"
+	"code.gitea.io/gitea/modules/util"
+
+	licenseclassifier "github.com/google/licenseclassifier/v2"
+)
+
+var (
+	classifier     *licenseclassifier.Classifier
+	licenseAliases map[string]string
+
+	// licenseUpdaterQueue represents a queue to handle update repo licenses
+	licenseUpdaterQueue *queue.WorkerPoolQueue[*LicenseUpdaterOptions]
+)
+
+func AddRepoToLicenseUpdaterQueue(opts *LicenseUpdaterOptions) error {
+	if opts == nil {
+		return nil
+	}
+	return licenseUpdaterQueue.Push(opts)
+}
+
+func loadLicenseAliases() error {
+	if licenseAliases != nil {
+		return nil
+	}
+
+	data, err := options.AssetFS().ReadFile("", "license-aliases.json")
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal(data, &licenseAliases)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func ConvertLicenseName(name string) string {
+	if err := loadLicenseAliases(); err != nil {
+		return name
+	}
+
+	v, ok := licenseAliases[name]
+	if ok {
+		return v
+	}
+	return name
+}
+
+func initClassifier() error {
+	if classifier != nil {
+		return nil
+	}
+
+	// threshold should be 0.84~0.86 or the test will be failed
+	classifier = licenseclassifier.NewClassifier(.85)
```

Comment:
This initClassifier causes data-race problem because you call it in many uncontrolled places.

For example, in detectLicense, it might be called when 2 repos are being pushed.

Then:

1. data race (reading/write to classifier is not protected)
2. dirty init: when in one go-routine `classifier` becomes non-nil, it is still not ready, another go-routine will use a dirty classifier

Reply:
2403c8085897ab9884bcc26632fe12e60197a35d